### PR TITLE
Adding preMaterialize operator for Source #23895

### PR DESCRIPTION
**Pull request by svezfaz**
Refs #23895

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
PLS WHITELIST

**Comment by svezfaz** [comment]
Missed an import in the docs test. Should be passing now.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala [hunk 1]
```diff
@@ -384,4 +384,61 @@ class SourceSpec extends StreamSpec with DefaultTimeout {
...
+
+      mat.fail(new RuntimeException("boom"))
+
+      probe1.expectSubscription()
+      probe2.expectSubscription()
+
+      probe1.expectError().getMessage should ===("boom")
+      probe2.expectError().getMessage should ===("boom")
+    }
+  }
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala]: Maybe add one test case for when materialization of the matVal powered source fails also (`.mapMaterializedValue(_ => throw booom)`), mostly to document what happens.

**Review by johanandren — APPROVED** [review]
LGTM with some minor comments

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 2]
```diff
@@ -87,6 +87,15 @@ final class Source[+Out, +Mat](
   override def mapMaterializedValue[Mat2](f: Mat ⇒ Mat2): ReprMat[Out, Mat2] =
     new Source[Out, Mat2](traversalBuilder.transformMat(f.asInstanceOf[Any ⇒ Any]), shape)
 
+  /**
+   * Materializes this Source, immediately returning (1) its materialized value, and (2) a new Source
+   * that can be used to consume elements from the newly materialized Source.
+   */
+  def preMaterialize()(implicit materializer: Materializer): (Mat, ReprMat[Out, NotUsed]) = {
+    val (mat, pub) = toMat(Sink.asPublisher(fanout = true))(Keep.both).run()
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: good, let's keep it like that. Technically the end-user may want to decide that but let's not expose this decision until we actually go into implementing the real-thing, as then perhaps that would not be needed at all.

**Review by ktoso — APPROVED** [review]

**Pull request merged**